### PR TITLE
Dynamic tenancy configurations

### PR DESCRIPTION
**Pull request by abhivka7**
### Description
* Category : Enhancement
* Why these changes are required?
These changes are required to dynamically enable/disable multi-tenancy, private tenant and set a default tenant.
* What is the old behavior before changes and new behavior after changes?
Earlier users had to change YAML files in each node and restart node to enable/disable multi-tenancy and private tenant. With our changes, users will be able to change these settings dynamically. We will also give admins the option to set a default tenant for all users for first time log in. This PR is for changes in Dashboards so that users can use Dashboards UI to change these settings. 
Related Security plugin PR: https://github.com/opensearch-project/security/pull/2607
More details can be found here:
https://github.com/opensearch-project/OpenSearch/issues/5853
https://github.com/opensearch-project/security-dashboards-plugin/issues/1302

### Issues Resolved
https://github.com/opensearch-project/security-dashboards-plugin/issues/1302

Is this a backport? If so, please add backport PR # and/or commits #
Nope

### Testing
We have done manual testing for all changes on our local server. We will also write Integ tests for these changes.

### Check List
- [  ] New functionality includes testing
- [] New functionality has been documented. : Will document in further commits.
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

**This is still a draft request. I am still working on some minor fixes and code style.**

### CODE REVIEW

File: public/apps/configuration/panels/tenant-list/save_changes_modal.tsx [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  }
+
+  if((props.updatedTenancyConfig.private_tenant_enabled!=props.originalTenancyConfig.private_tenant_enabled) && props.updatedTenancyConfig.multitenancy_enabled){
+    privateTenantToggled = true;
+    if(props.updatedTenancyConfig.private_tenant_enabled) {
+      privateTenancyChangeMessage="Enable private tenant - Users will be able to create index patterns, visualizations, and" +
+        "other OpenSearch Dashboards objects in a private tenant that they only have access to."
+    }
+    else{
+      privateTenancyChangeMessage="Disabling private tenant - Users won't be able to access index patterns, visualizations, and" +
```
Comment (scrawfor99) [inline public/apps/configuration/panels/tenant-list/save_changes_modal.tsx:111]: Maybe I am reading this wrong, but would this say the "Disabling..." message even if `updatedTenancy` was `private_tenant_enabled` and `origingTenancy` was `private_tenant_enabled`? Is it possible to reach here without the config update being different? I.e. someone sets the config to what it already is?

File: public/apps/configuration/utils/request-utils.ts [hunk 2]
```diff
@@ -30,6 +30,10 @@ export async function httpPost<T>(http: HttpStart, url: string, body?: object):
   return await request<T>(http.post, url, body);
 }
 
+export async function httpPut<T>(http: HttpStart, url: string, body?: object): Promise<T> {
+  return await request<T>(http.put, url, body);
```
Comment (scrawfor99) [inline public/apps/configuration/utils/request-utils.ts:34]: If body can be optional is this going to throw an error when it is null?

File: public/apps/configuration/utils/tenant-utils.tsx [hunk 3]
```diff
@@ -78,15 +81,21 @@ export function transformTenantData(
...
-      tenantValue: PRIVATE_USER_DICT.Value,
-    });
-  }
+  tenantList.splice(1, 0, {
+    tenant: PRIVATE_USER_DICT.Label,
+    reserved: true,
+    description: PRIVATE_USER_DICT.Description,
+    tenantValue: PRIVATE_USER_DICT.Value,
+  });
+  // if (isPrivateEnabled) {
```
Comment (scrawfor99) [inline public/apps/configuration/utils/tenant-utils.tsx:90]: Looks like this still has not gotten removed yet.

File: public/apps/configuration/utils/toast-utils.tsx [hunk 4]
```diff
@@ -33,6 +44,22 @@ export function createUnknownErrorToast(id: string, failedAction: string): Toast
...
 
+export function createTenancyErrorToast(id: string, Title:string, Message: string): Toast {
+  return createErrorToast(
+    id,
+    Title,
+    Message
+  );
+}
+
+export function createTenancySuccessToast(id: string, Title:string, Message: string): Toast {
```
Comment (scrawfor99) [inline public/apps/configuration/utils/toast-utils.tsx:55]: Nit: For parameter consistency consider swapping to `title` (lowercase)

File: public/apps/configuration/utils/toast-utils.tsx [hunk 5]
```diff
@@ -33,6 +44,22 @@ export function createUnknownErrorToast(id: string, failedAction: string): Toast
   );
 }
 
+export function createTenancyErrorToast(id: string, Title:string, Message: string): Toast {
```
Comment (scrawfor99) [inline public/apps/configuration/utils/toast-utils.tsx:47]: Nit: For parameter consistency consider swapping to `title` (lowercase)

File: public/types.ts [hunk 6]
```diff
@@ -34,12 +34,18 @@ export interface SecurityPluginStartDependencies {
 }
 
 export interface AuthInfo {
-  user_name: string;
-  tenants: {
+  user_name?: string;
```
Comment (scrawfor99) [inline public/types.ts:37]: Why is the optional now?

File: server/auth/types/authentication_type.ts [hunk 7]
```diff
@@ -246,15 +246,33 @@ export abstract class AuthenticationType implements IAuthenticationType {
...
     }
 
-    const selectedTenant = resolveTenant(
-      request,
-      authInfo.user_name,
-      authInfo.roles,
-      authInfo.tenants,
-      this.config,
-      cookie
+    // return resolveTenant(
```
Comment (scrawfor99) [inline server/auth/types/authentication_type.ts:249]: Looks like this can be removed?

File: server/auth/types/authentication_type.ts [hunk 8]
```diff
@@ -246,15 +246,33 @@ export abstract class AuthenticationType implements IAuthenticationType {
...
+        config: this.config,
+        cookie: cookie,
+        multitenancy_enabled: authInfo.multitenancy_enabled,
+        private_tenant_enabled: authInfo.private_tenant_enabled,
+        default_tenant: authInfo.default_tenant
+      }
     );
-    return selectedTenant;
+
+    // return selectedTenant;
```
Comment (scrawfor99) [inline server/auth/types/authentication_type.ts:275]: Return is commented out--is this supposed to be un-commented?

File: server/auth/types/basic/routes.ts [hunk 9]
```diff
@@ -203,14 +209,19 @@ export class BasicAuthRoutes {
             expiryTime: Date.now() + this.config.session.ttl,
           };
 
-          if (this.config.multitenancy?.enabled) {
+          if (user.multitenancy_enabled) {
```
Comment (scrawfor99) [inline server/auth/types/basic/routes.ts:212]: Below you have `multitenancy_enabled?` but here it is not optional. Is this intentional? Just want to make sure you meant them to be different.

**Review by scrawfor99 — COMMENTED** [review]
Hi @abhivka7, thank you for your incredibly detailed PR. This looks really good and it is clear how much time you have put into polishing things. I left some comments with a few questions and a few nitpicks. 

The only thing I saw that would be required for me to approve would be the addition of some tests to exercise the changes you made. Since you have numerous UX/UI sided elements, you should be able to test your functionality with Cypress. There are some existing tests that you can likely use as a template. I would be interested to see confirmation that the tenancy changes do take effect after the dynamic swap and also be curious if you are able to test the configuration of the default tenant behavior you added. 

Thank you again and nice work :)

File: public/apps/configuration/utils/tenant-utils.tsx [hunk 10]
```diff
@@ -34,18 +34,21 @@ import {
...
   DEFAULT_TENANT,
   GLOBAL_TENANT_RENDERING_TEXT,
   GLOBAL_TENANT_SYMBOL,
   globalTenantName,
   isGlobalTenant,
   isRenderingPrivateTenant,
-  PRIVATE_TENANT_RENDERING_TEXT,
+  PRIVATE_TENANT_RENDERING_TEXT, SAML_AUTH_LOGIN,
 } from '../../../../common';
+import {TenancyConfigSettings} from '../panels/tenancy-config/types';
```
Comment (cliu123) [inline public/apps/configuration/utils/tenant-utils.tsx:51]:
```suggestion
import { TenancyConfigSettings } from '../panels/tenancy-config/types';
```

File: public/apps/configuration/utils/tenant-utils.tsx [hunk 11]
```diff
@@ -69,7 +72,7 @@ export async function fetchTenantNameList(http: HttpStart): Promise<string[]> {
 
 export function transformTenantData(
   rawTenantData: DataObject<Tenant>,
-  isPrivateEnabled: boolean
+  isPrivateEnabled?: boolean
```
Comment (cliu123) [inline public/apps/configuration/utils/tenant-utils.tsx:75]: Please remove this argument if no longer needed.

File: public/apps/configuration/utils/tenant-utils.tsx (hunk 3 above)
Comment (cliu123) [reply public/apps/configuration/utils/tenant-utils.tsx:90]: @abhivka7 Why adding private tenant to the list anyways? Shouldn't it be added only when private tenant is enabled?

File: public/plugin.ts [hunk 12]
```diff
@@ -42,11 +42,12 @@ import {
   ClientConfigType,
   SecurityPluginSetup,
   SecurityPluginStart,
-  SecurityPluginSetupDependencies,
+  SecurityPluginSetupDependencies, AuthInfo,
```
Comment (cliu123) [inline public/plugin.ts:45]: Please remove unnecessary changes.

File: public/plugin.ts [hunk 13]
```diff
@@ -42,11 +42,12 @@ import {
...
 } from './types';
 import { addTenantToShareURL } from './services/shared-link';
 import { interceptError } from './utils/logout-utils';
 import { tenantColumn, getNamespacesToRegister } from './apps/configuration/utils/tenant-utils';
+import { getAuthInfo } from './utils/auth-info-utils';
```
Comment (cliu123) [inline public/plugin.ts:50]: Please remove unnecessary changes.

File: public/plugin.ts [hunk 14]
```diff
@@ -101,14 +103,17 @@ export class SecurityPlugin
           const { renderApp } = await import('./apps/configuration/configuration-app');
           const [coreStart, depsStart] = await core.getStartServices();
 
+          // const auth1 = (await getAuthInfo(props.coreStart.http))
+
```
Comment (cliu123) [inline public/plugin.ts:107]: Please remove unnecessary changes.

File: server/auth/user.ts [hunk 15]
```diff
@@ -21,4 +21,7 @@ export interface User {
   selectedTenant?: string;
   credentials?: any;
   proxyCredentials?: any;
+  multitenancy_enabled?:boolean;
+  private_tenant_enabled?:boolean;
+  default_tenant?: string;
```
Comment (cliu123) [inline server/auth/user.ts:26]: Aren't these settings on node/cluster level? Why are they added to `User` interface?

File: public/types.ts (hunk 6 above)
Comment (abhivka7) [reply public/types.ts:37]: Done while testing something, removing it now. Thnaks.

File: public/apps/configuration/configuration-app.tsx [hunk 16]
```diff
@@ -19,16 +19,17 @@ import React from 'react';
...
 export function renderApp(
   coreStart: CoreStart,
   navigation: SecurityPluginStartDependencies,
   params: AppMountParameters,
-  config: ClientConfigType
+  config: ClientConfigType,
+  dashboardsInfo: DashboardsInfo
 ) {
-  const deps = { coreStart, navigation, params, config };
+  const deps = { coreStart, navigation, params, config, dashboardsInfo };
```
[the remark this answers is not in the document]
Comment (abhivka7) [reply public/apps/configuration/configuration-app.tsx:32]: Yup added while testing. Removed now.

File: public/apps/configuration/utils/tenant-utils.tsx (hunk 11 above)
Comment (abhivka7) [reply public/apps/configuration/utils/tenant-utils.tsx:75]: Cool. Thanks.

File: public/apps/configuration/utils/tenant-utils.tsx (hunk 3 above)
Comment (abhivka7) [reply public/apps/configuration/utils/tenant-utils.tsx:90]: @cliu123 With the new UI design changes, we want to always keep private tenant in list. For manage tab, we will have a new column called Status which shows when a tenant is disabled. Removing private tenant from the list cause confusion among users and this new approach of Status column is more descriptive.

Thanks.

File: public/types.ts (hunk 6 above)
Comment (abhivka7) [reply public/types.ts:37]: Was done while testing, fixing now.

File: server/auth/types/authentication_type.ts (hunk 8 above)
Comment (abhivka7) [reply server/auth/types/authentication_type.ts:275]: Nope. Removing the line.

File: server/auth/types/basic/routes.ts (hunk 9 above)
Comment (abhivka7) [reply server/auth/types/basic/routes.ts:212]: Yes

File: server/auth/user.ts (hunk 15 above)
Comment (abhivka7) [reply server/auth/user.ts:26]: Going forward, this will be useful if we might want to make these settings user level. That's why keeping it here.

**Comment by abhivka7** [comment]
@scrawfor99 @cliu123 I have added cypress tests for these changes here: https://github.com/opensearch-project/opensearch-dashboards-functional-test/pull/609/files. 
Please take a look.

File: server/auth/user.ts (hunk 15 above)
Comment (cliu123) [reply server/auth/user.ts:26]: @abhivka7 How would it work if we make these settings on user level? Would it only support internal users? What about external users such as SAML users?

File: .github/workflows/cypress-test.yml [hunk 17]
```diff
@@ -70,3 +70,5 @@ jobs:
           cd opensearch-dashboards-functional-test
           npm install cypress --save-dev
           yarn cypress:run-with-security-and-aggregation-view --browser chrome --spec "cypress/integration/plugins/security-dashboards-plugin/aggregation_view.js"
+          yarn cypress:run-with-security --browser chrome --spec "cypress/integration/plugins/security-dashboards-plugin/multi_tenancy.js"
```
Comment (RyanL1997) [inline .github/workflows/cypress-test.yml:73]: Since the directory of these tests are the same, I think you can just replace the original command line with`yarn cypress:run-with-security --browser chrome --spec "cypress/integration/plugins/security-dashboards-plugin/*.js"`

**Review by RyanL1997 — COMMENTED** [review]
Hi @abhivka7 Thanks for putting these together! I see you have fixed the cypress issue! I think for now, this is good to be merged if we fixed the integration test failure + unit test failures introduced by some UI changes.

File: .github/workflows/cypress-test.yml (hunk 17 above)
Comment (RyanL1997) [reply .github/workflows/cypress-test.yml:73]: Oh nvm, I think some of the flag are not the same, cuz your cases are `run-with-security`. So lets just leave like this.

File: test/jest_integration/saml_auth.test.ts [hunk 18]
```diff
@@ -284,87 +284,87 @@ describe('start OpenSearch Dashboards server', () => {
     await driver.quit();
   });
 
-  it('Tenancy persisted after Logout in SAML', async () => {
-    const driver = getDriver(browser, options).build();
```
Comment (abhivka7) [inline test/jest_integration/saml_auth.test.ts:288]: After discussion with @devardee, we concluded that this test is flaky and needs to be fixed. I will open a github issue soon for this test.
cc: @anijain-Amazon @prabhat-chaturvedi

File: test/jest_integration/saml_auth.test.ts [hunk 19]
```diff
@@ -284,87 +284,87 @@ describe('start OpenSearch Dashboards server', () => {
...
+  //     expect(newtenantName).toEqual('Global');
+  //   } else {
+  //     expect(localStorageItem).toEqual('""');
+  //     expect(tenantName).toEqual('Global');
+  //   }
+  //   await driver.manage().deleteAllCookies();
+  //   await driver.quit();
+  //
+  //   expect(localStorageItem).toEqual('""');
+  // });
```
Comment (RyanL1997) [inline test/jest_integration/saml_auth.test.ts:367]: I think you dont want to comment the saml testing out. I can re-run it if it fails

File: public/apps/configuration/app-router.tsx [hunk 20]
```diff
@@ -232,7 +241,14 @@ export function AppRouter(props: AppDependencies) {
               path={ROUTE_MAP.tenants.href}
               render={() => {
                 setGlobalBreadcrumbs(ResourceType.tenants);
-                return <TenantList {...props} />;
+                return <TenantList tabID={'Manage'} {...props} />;
```
Comment (cwperks) [inline public/apps/configuration/app-router.tsx:244]: How will i18n work here?

Comment (cwperks) [reply public/apps/configuration/app-router.tsx:244]: You can look at usages of `translateMessage` for how to do i18n in this repo: https://github.com/opensearch-project/security-dashboards-plugin/blob/main/public/apps/translation-utils.ts#L18-L20

File: server/auth/user.ts (hunk 15 above)
Comment (cwperks) [reply server/auth/user.ts:26]: Wouldn't `default_tenant` be the only one of these settings that may be applicable on the user-level? How would it work if some users had multitenancy_enabled, but others don't?

File: public/apps/configuration/utils/tenant-utils.tsx [hunk 21]
```diff
@@ -67,26 +71,20 @@ export async function fetchTenantNameList(http: HttpStart): Promise<string[]> {
   return Object.keys(await fetchTenants(http));
 }
 
-export function transformTenantData(
-  rawTenantData: DataObject<Tenant>,
-  isPrivateEnabled: boolean
-): Tenant[] {
+export function transformTenantData(rawTenantData: DataObject<Tenant>): Tenant[] {
```
Comment (scrawfor99) [inline public/apps/configuration/utils/tenant-utils.tsx:74]: Why was the `isPrivateEnabled` field removed here?

File: public/apps/configuration/utils/toast-utils.tsx [hunk 22]
```diff
@@ -33,6 +44,14 @@ export function createUnknownErrorToast(id: string, failedAction: string): Toast
   );
 }
 
+export function createTenancyErrorToast(id: string, title: string, Message: string): Toast {
+  return createErrorToast(id, title, Message);
+}
+
+export function createTenancySuccessToast(id: string, Title: string, Message: string): Toast {
+  return createSuccessToast(id, Title, Message);
```
Comment (scrawfor99) [inline public/apps/configuration/utils/toast-utils.tsx:52]: Can you just use the same function for both error and success.

File: public/apps/configuration/app-router.tsx (hunk 20 above)
Comment (abhivka7) [reply public/apps/configuration/app-router.tsx:244]: @cwperks We need to pass tabID in order to switch tabs.

Comment (abhivka7) [reply public/apps/configuration/app-router.tsx:244]: It can not be a constant as we want to call TenantList with different tabs based on our use case.

File: public/apps/configuration/panels/tenant-list/test/__snapshots__/tenant-list.test.tsx.snap [hunk 23]
```diff
@@ -82,6 +71,14 @@ exports[`Tenant list Action menu click Duplicate click 1`] = `
             >
               Create visualizations
             </EuiButtonEmpty>
+            <EuiButtonEmpty
+              disabled={true}
+              id="set as Global Default"
```
Comment (cwperks) [inline public/apps/configuration/panels/tenant-list/test/__snapshots__/tenant-list.test.tsx.snap:76]: Nit: avoid spaces in ids and class names. HTML does not allow spaces in `id` because it cannot be referenced by css. It looks like it works in React, but IMO spaces should be avoided.

Comment (abhivka7) [reply public/apps/configuration/panels/tenant-list/test/__snapshots__/tenant-list.test.tsx.snap:76]: I can take this up later to fix this.

**Review by cwperks — APPROVED** [review]
Thank you for a great contribution @abhivka7! I'm approving this PR, but also have an open question on tying `multitenancy_enabled` to a user. Can you take a look at the comment when you have a moment? https://github.com/opensearch-project/security-dashboards-plugin/pull/1394/files#r1169156020

**Review by scrawfor99 — APPROVED** [review]
Hi @abhivka7, thank you for all your work on this PR. I left some comments, but I did not see anything that I would hold up the PR for.

File: public/apps/configuration/app-router.tsx [hunk 24]
```diff
@@ -62,6 +62,14 @@ const ROUTE_MAP: { [key: string]: RouteItem } = {
     name: 'Tenants',
     href: buildUrl(ResourceType.tenants),
   },
+  [ResourceType.tenantsManageTab]: {
+    name: 'TenantsManageTab',
+    href: buildUrl(ResourceType.tenantsManageTab),
+  },
+  [ResourceType.tenantsConfigureTab]: {
+    name: '',
```
[the remark this answers is not in the document]
Comment (abhivka7) [reply public/apps/configuration/app-router.tsx:70]: Nope on purpose.

**Comment by andreasMore** [comment]
I would like to deactivate the private tenant in general. To do this, I have added the option "opensearch_security.multitenancy.tenants.enable_private: false" to "opensearch_dashboards.yml". Unfortunately, it is still possible to select the private tenant afterwards. Is it now desired that a change can only be made through the UI?

In the file "tenant-switch-panel" the state "isPrivateEnabled" is set to the config props of the opensearch_dashboards.yml. Further down in the file, this state is then overwritten with the fetched data. Therefore, it is currently not possible to disable the private tenant via the "opensearch_dashboards.yml".

**Comment by abhivka7** [comment]
It is suggested to use either the UI or the APIs to configure tenancy properties.
You can use this API:
```
PUT _plugins/_security/api/tenancy/config 
{
    "private_tenant_enabled":false
}
```
to disable private tenant if you don't want to go through UI.

For further details refer here: https://aws.amazon.com/blogs/big-data/configure-dynamic-tenancy-for-amazon-opensearch-dashboards/
